include/seqan3/range/view/single_pass_input.hpp

```diff
@@ -0,0 +1,402 @@
+    ~single_pass_input_view() = default;
+
+    //!\brief Construction from the underlying view.
+    single_pass_input_view(urng_t && urng) :
+        state_ptr{new state{std::forward<urng_t>(urng), urng.begin()}}
+    {}
+    //!\}
+
+    /*!\name Iterators
+     * \{
+     */
+    /*!\brief Returns an iterator to the current begin of the underlying range.
+     *
+     * \details
+     *
+     * Subsequent calls to begin will result in different positions if the iterator was incremented
+     * between the calls.
+     */
+    iterator begin()
+    {
+        return iterator{*this};
+    }
+
+    //!\brief Const version of begin is deleted, since the underlying state must be mutable.
+    const_iterator begin() const = delete;
+
+    //!\copydoc single_pass_input_view::begin() const
+    const_iterator cbegin() const = delete;
+
+    //!\brief Returns a sentinel.
+    sentinel end()
+    {
+        // calling to urng->end must be in O(1).
+        return sentinel{state_ptr->urng.end()};
+    }
+
+    //!\brief Const version of end is deleted, since the underlying state must be mutable.
+    sentinel end() const = delete;
+
+    //!\copydoc single_pass_input_view::end() const
+    sentinel cend() const = delete;
+    //!\}
+};
+
+/*!\name Deduction guide.
+ * \relates seqan3::detail::single_pass_input_view
+ * \{
+ */
+//!\brief Deduces the single_pass_input_view from the underlying range.
+template <input_range_concept urng_t>
+single_pass_input_view(urng_t &&) -> single_pass_input_view<urng_t>;
+//!\}
+} // seqan3::detail
+
+//-----------------------------------------------------------------------------
+// Iterator for single pass input view.
+//-----------------------------------------------------------------------------
+
+namespace seqan3::detail
+{
+/*!\brief An input_iterator over the associated range.
+ * \implements input_iterator_concept
+ * \ingroup view
+ * \tparam view_type The type of the associated type.
+ *
+ * This iterator reduces every iterator type of the associated view to an single pass input iterator.
+ */
+template <typename view_type>
+class single_pass_input_iterator<single_pass_input_view<view_type>>
+{
+    //!\brief The pointer to the associated view.
+    using base_iterator_t = typename single_pass_input_view<view_type>::urng_iterator_type;
+    //!\brief The pointer to the associated view.
+    single_pass_input_view<view_type> * view_ptr{};
+
+    //!\ Friend declaration to give seqan3::detail::single_pass_input_sentinel access to members of this class.
+    template <typename sentinel_t>
+    friend class single_pass_input_sentinel;
+
+public:
+
+    /*!\name Member types
+     * \{
+     */
+    //!\brief Difference type.
+    using difference_type   = difference_type_t<base_iterator_t>;
+    //!\brief Value type.
+    using value_type        = value_type_t<base_iterator_t>;
+    //!\brief Pointer type.
+    using pointer           = std::add_pointer_t<value_type>;
```

Comment:
pointer usually is the iterator type, I think, but it always confused me :thinking:

Reply:
In any case it is wrong here. It should be simply the pointer type of the underlying iterator. 
Given that this can also be an output iterator. I'll think about it.